aidial_assistant/model/model_client.py

```diff
@@ -70,43 +76,53 @@ async def agenerate(
         extra_results_callback: ExtraResultsCallback | None = None,
         **kwargs,
     ) -> AsyncIterator[str]:
-        async with ClientSession(read_bufsize=self.buffer_size) as session:
-            openai.aiosession.set(session)
-
-            model_result = await openai.ChatCompletion.acreate(
-                messages=[message.to_openai_message() for message in messages],
-                **self.model_args | kwargs,
-            )
-
-            finish_reason_length = False
-            async for chunk in model_result:  # type: ignore
-                usage: Usage | None = chunk.get("usage")
-                if usage:
-                    prompt_tokens = usage["prompt_tokens"]
-                    self._total_prompt_tokens += prompt_tokens
-                    self._total_completion_tokens += usage["completion_tokens"]
-                    if extra_results_callback:
-                        extra_results_callback.on_prompt_tokens(prompt_tokens)
-
+        model_result = await self.client.chat.completions.create(
+            **self.model_args,
+            extra_body=kwargs,
+            stream=True,
+            messages=[message.to_openai_message() for message in messages],  # type: ignore
+        )  # type: ignore
+
+        finish_reason_length = False
+        tool_calls_chunks = []
+        async for chunk in model_result:  # type: ignore
+            chunk = chunk.dict()
```

Comment:
1. `chunk` has type `ChatCompletionChunk`, so you could access `chunk.usage` and `chunk.choices` in a type-safe manner.
2. `ChatCompletionChunk` is instance of `pydantic.BaseModel`, meaning that `openai` SDK validates the chat completion responses. So the response with DIAL-specific fields like `statistics` might be rejected. Please double check they are not rejected. If they are, then we have to avoid using `openai` SDK.

Reply:
The ChatCompletionChunk doesn't have usage field. It comes as an extra field.

Reply:
1. Still holds for `chunk.choices`
2. Case closed: `openai.BaseModel` is defined as `pydantic.BaseModel` with `allow_extra=True`, so we don't lost any extra DIAL fields.

Reply:
Using the typed object now.